salt/modules/cmdmod.py

```diff
@@ -407,16 +407,27 @@ def _get_stripped(cmd):
         return win_runas(cmd, runas, password, cwd)
 
     if runas and salt.utils.platform.is_darwin():
-        # we need to insert the user simulation into the command itself and not
-        # just run it from the environment on macOS as that
-        # method doesn't work properly when run as root for certain commands.
+        # We need to insert the user simulation into the command itself and not
+        # just run it from the environment on macOS as that method doesn't work
+        # properly when run as root for certain commands.
         if isinstance(cmd, (list, tuple)):
             cmd = ' '.join(map(_cmd_quote, cmd))
 
-        cmd = 'su -l {0} -c "{1}"'.format(runas, cmd)
-        # set runas to None, because if you try to run `su -l` as well as
-        # simulate the environment macOS will prompt for the password of the
-        # user and will cause salt to hang.
+        # Ensure directory is correct before running command
+        cmd = 'cd -- {dir} && {{ {cmd}\n }}'.format(dir=_cmd_quote(cwd), cmd=cmd)
+
+        # Ensure environment is correct for a newly logged-in user by running
+        # the command under bash as a login shell
+        cmd = '/bin/bash -l -c {cmd}'.format(cmd=_cmd_quote(cmd))
```

Comment:
I think this could be worked around by looking at `__salt__['user.info'](runas)['shell']`, though we may then run into a problem if we are trying to run a command as a user with `/usr/bin/false` or `/sbin/nologin` as their shell, or if they have a shell like `fish` which accepts a different syntax.

For the record, my main goal in using `bash -l` is to get the environment into a better state than the one `su -l` leaves it in (which doesn't have /usr/local/bin in the path, for example), so I still consider this a fair improvement over the previous behaviour.

Would it be best to launch the command under bash/zsh depending on macos version, or under the user's default shell, or maybe is there a third option I haven't thought of?

Reply:
Perhaps that mystery third option is to use zsh when the shell is zsh, and bash otherwise...

Reply:
It seems like `zsh` is working well just with `su -l` as you can see in my example above where the `PATH` is properly set and all my environment variables from `.zprofile` are available.

So maybe it is only necessary to check wether the user's default shell is `bash` and only if it is `bash` then add that line.

Also, an improvement for that line is to call the right bash executable since it could be `/usr/local/bin/bash`